### ISSUE TITLE
IMPROVEMENT: Better handle timestamps (bucket)

### ISSUE THREAD

**Issue by CraigHawker** [M-Files/VAF.Extensions.Community#112]
There are a few issues when using timestamps and dotnet DateTime instances:

- When retrieving a TIMESTAMP property value (e.g. via [ObjVerEx.GetPropertyAs<DateTime?>](https://github.com/M-Files/VAF.Extensions.Community/blob/616a574d0e8da332d1a3c71e1476550597e32787/MFiles.VAF.Extensions/ExtensionMethods/ObjVerEx/GetPropertyAs.cs#L74)) the value is in UTC, but DateTimeKind is set to Unspecified.  This can cause odd behaviour in consuming applications and may be the root cause of https://github.com/M-Files/VAF.Extensions.Community/issues/87.
- [ObjVerEx.GetPropertyAs<DateTime?>()](https://github.com/M-Files/VAF.Extensions.Community/blob/616a574d0e8da332d1a3c71e1476550597e32787/MFiles.VAF.Extensions/ExtensionMethods/ObjVerEx/GetPropertyAs.cs#L74) will lose fractional seconds when used with TIMESTAMP property values.  This can cause some issues with searches, in particular.  We should use `TypedValue.GetValueAsTimestamp().ToDateTime()` to fix this.
- We should add `MFSearchBuilder.Timestamp()` which will do the same as the [Date](https://github.com/M-Files/VAF.Extensions.Community/blob/616a574d0e8da332d1a3c71e1476550597e32787/MFiles.VAF.Extensions/ExtensionMethods/MFSearchBuilderExtensionMethods/DateTimes.cs#L108) method but incorporates time.  The API expects the value to be in local time.  If we pass in a DateTime whose kind is set to UTC then we should adjust for the offset.  May also need to use `DateTime.ToTimestamp` to include fractional seconds (?).
- Some helpers for "before/after/between" would be nice for helpers when working with dates/times/timestamps.

**Comment by CraigHawker** [comment]
Potential helper methods, from Joel:

```csharp
        public static Timestamp ToPreciseTimestamp( this DateTime dateTime )
        {
            return new TimestampClass
            {
                Year = ( uint )dateTime.Year,
                Month = ( uint )dateTime.Month,
                Day = ( uint )dateTime.Day,
                Hour = ( uint )dateTime.Hour,
                Minute = ( uint )dateTime.Minute,
                Second = ( uint )dateTime.Second,
                Fraction = ( uint )( dateTime.Ticks % 1e7M * 100 )  // 10,000,000 ticks in a second, 100 nanoseconds in a tick
            };
        }

        public static DateTime ToPreciseDateTime(
            this Timestamp timestamp,
            DateTimeKind kind = DateTimeKind.Local
        )
        {
            return new DateTime( 
                ( int )timestamp.Year,
                ( int )timestamp.Month,
                ( int )timestamp.Day,
                ( int )timestamp.Hour,
                ( int )timestamp.Minute,
                ( int )timestamp.Second,
                kind
                ).AddTicks( timestamp.Fraction / 100 );   // 100 nanoseconds in a tick.
        }
```

**Comment by CraigHawker** [comment]
Preview available as [23.8.0.73-timestamp-search-conditions](https://www.nuget.org/packages/MFiles.VAF.Extensions/23.8.0.73-timestamp-search-conditions).
